Proposals/20200511-HVPA-design.md

```diff
@@ -0,0 +1,235 @@
+
+**Note:** Renaming the fields `StartReplicaCount` and `LastReplicaCount` to `minReplicas` and `maxReplicas` respectively.
+
+Example 1:
+```yaml
+        -   minReplicas:  1,
+            maxReplicas:  1,
+            vpaWeight:    100,
+            maxCPU:       2,
+            maxMemory:    "4G",
+
+        -   minReplicas:  2,
+            maxReplicas:  4,
+            vpaWeight:    50,
+
+        -   minReplicas:  5,
+            maxReplicas:  5,
+            vpaWeight:    100,
+            minCPU:       5,
+            minMemory:    "10G",
+```
+Current replica count is 1, and resources are 1.5 cores / 2.5GB. If VPA recommends 2.5 cores and 4.5GB, and HPA recommends 2 replicas, then the HVPA controller would scale the application as following:
+* New replica count: 2 (ceil of 50% of change recommended by HPA is applied)
+* New requests: 2 cores, 3.5GB (only 50% of change recommended by VPA is applied)
+
+Current replica count is 5, and resources are 6 cores / 12GB. If VPA recommends 3000m CPU and 3GB, and HPA recommends 4 replicas, the HVPA controller would scale the application as following:
+* New replica count: 4 (floor of 50% of change recommended by HPA is applied)
+* New requests: 4.5 cores, 7.5GB (only 50% of VPA changes are applied)
+
+Example 2:
+```yaml
+        -   minReplicas:  1,
+            maxReplicas:  1,
+            VpaWeight:    100,
+            maxCPU:       2,
+            maxMemory:    "4G",
+
+        -   minReplicas:  1,
+            maxReplicas:  1,
+            VpaWeight:    0,
+
+        -   minReplicas:  2,
+            maxReplicas:  2,
+            VpaWeight:    100,
+            minCPU:       1.5,
+            minMemory:    "3.5G",
+            maxCPU:       5,
+            maxMemory:    "8G",
+
+        -   minReplicas:  2,
+            maxReplicas:  2,
+            VpaWeight:    0,
+```
+This kind of spec can be used to enable only vertical scaling at each replica count with some minimum and maximum replica size configured for that replica count. If the VPA recommendations are out of the range specified in the "current" bucket, then HVPA controller would scale the application horizontally as per rules described above.
+
+**Pod with multiple containers**
+
+In such a case, all the calculations would be done using the total CPU and total memory allocated to all the containers in the pod for which VPA container resource policy is set to "Auto". The total weighted CPU and memory per replica would then be divided in the same proportion as the recommendations by VPA for those containers.
+
+### Pros:
+Quick to implement as it builds upon existing implementation.
+
+### Cons:
+HVPA controller takes `ceil` of weighted replica count when scaling. If HPA recommends in increments of 1, then HVPA controller would end up taking full HPA recommendations, and also have a partial vertical scaling. This results in overscaling especially if HPA recommends in increments of 1.
```

Comment:
- `ceil` is not the problem here. The problem is inherent in the fact that replicas have to be integers.
- Overscaling is not the problem either. Even the current proposal has overscaling problem (now vertically)  while switching buckets.
- Also, it is not just this. For example, the decision to use current replicas to decide the buckets is also ad hoc with other equally good/bad alternatives with the same problems.

The real reason behind the proposal is to gain a bit more control over how the scaling happens vertically to make sure the correlation between the number of replicas and the vertical scale of the replicas is more controlled with less variations in different cases.

Reply:
updated accordingly in https://github.com/gardener/hvpa-controller/commit/f7bcb341c6330ad0583a378dff390e753a6065ca